server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -330,6 +330,10 @@ public String getDescription() {
                     }
 
                     logger.debug("publishing cluster state version [{}]", newClusterState.version());
+                    // initialize routing nodes and the indices lookup concurrently, we will need both of them for the cluster state
+                    // application and can compute them while we wait for the other nodes during publication
+                    threadPool.generic().execute(newClusterState::getRoutingNodes);
+                    threadPool.generic().execute(() -> newClusterState.metadata().getIndicesLookup());
```

Comment:
I benchmarked this without the locking and with profiling and even at close to 100k indices these would always run before the applier thread gets to needing either of these two objects, so while not saving any cycles this at least makes better use of master's cores.

Reply:
Should we add some logging to find out when this starts to happen?

```
    private synchronized RoutingNodes buildRoutingNodes() {
        RoutingNodes r = routingNodes;
        if (r != null) {
            logger.warn("RoutingNodes object is already build.");
            return r;
        }
        r = RoutingNodes.immutable(routingTable, nodes);
        routingNodes = r;
        return r;
    }
```

Reply:
I don't think we can make this a warning, it will happen here and there for sure just by chance and that's ok IMO.
Even if this happens it's not so bad and still better than what we have today for heavy cases.
Either the applier thread enters the synchronized block before the generic threads (which is pretty unlikely) and the worst that happens is that we waste a generic thread blocked on nothing for a bit while it does the computation and performance is pretty much what we have right now.
Or the more likely case, the generic thread enters the block first happens, in which case even if it doesn't finish before the applier gets to the block, that's still better than the previous case of the applier having to do all the work itself, while now it merely has to wait for the already in-progress work on `GENERIC` to finish (though that seems to not be happening even at 200k indices in my benchmarks :)).